File: app/src/main/java/com/example/Sleepy/shared/MyTimer.java
package com.example.Sleepy.shared;

import android.content.Context;
import android.content.Intent;
import android.provider.AlarmClock;
import android.view.View;
import android.widget.TextView;
import android.widget.TimePicker;

import com.example.Sleepy.R;
import com.google.android.material.snackbar.Snackbar;

import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.util.Calendar;
import java.util.Date;
import java.util.Locale;
import java.util.Objects;

public class MyTimer {

    public static String calcRemainingTimeMinute(Date date, Context context){
        String RemTime;
        long timeUp = date.getTime();
        long diff = timeUp - new Date().getTime();

        long diffMinutes = diff / (60 * 1000) % 60;
        long diffHours = diff / (60 * 60 * 1000);

        if (diffMinutes == 0){
            RemTime = "" + diffHours + context.getString(R.string.hours);
        }else if(diffHours == 0){
            RemTime = "" + diffMinutes + context.getString(R.string.minute);
        }else{
            RemTime = "" + diffHours + context.getString(R.string.hours) + diffMinutes + context.getString(R.string.minute);
        }
        return RemTime;
    }

    public static String getFormatTime(long minute, Context context){
        String RemTime;
        long h, m;
        h = minute / 60;
        m = minute - h * 60;
        if (m == 0){
            RemTime = "" + h + context.getString(R.string.hours);
        }else if(h != 0 && m != 0){
            RemTime = "" + h + context.getString(R.string.hours) + m + context.getString(R.string.minute);
        }else{
            RemTime = "" + m + context.getString(R.string.minute);
        }
        return RemTime;
    }

    public static void clearTime(TimePicker tp, Context context){
        MyVibrator.vibrate(30, context);
        tp.setHour(Calendar.getInstance().get(Calendar.HOUR_OF_DAY));
        tp.setMinute(Calendar.getInstance().get(Calendar.MINUTE));
    }

    public static Calendar getCurrentTime(TimePicker tp){
        Calendar timeAlarm = Calendar.getInstance();
        timeAlarm.set(Calendar.MINUTE, tp.getMinute());
        timeAlarm.set(Calendar.HOUR_OF_DAY, tp.getHour());
        return timeAlarm;
    }

    public static void setAlarmInApp(Calendar time, Context context, View view){
        try{
            context.startActivity(new Intent(AlarmClock.ACTION_SET_ALARM)
                    .putExtra(AlarmClock.EXTRA_MESSAGE, "New Alarm")
                    .putExtra(AlarmClock.EXTRA_HOUR, time.get(Calendar.HOUR_OF_DAY))
                    .putExtra(AlarmClock.EXTRA_MINUTES, time.get(Calendar.MINUTE)));
        }catch (Exception ex){
            Snackbar.make(view, context.getString(R.string.error_open_alarm), Snackbar.LENGTH_LONG).show();
        }
    }

    public static void getAsleepText(int asleepTime, TextView tv, Context context) {
        if (asleepTime != 0){
            tv.setVisibility(View.VISIBLE);
            tv.setText(new StringBuilder()
                    .append(context.getString(R.string.calculate_with_sleep_time))
                    .append(asleepTime)
                    .append(context.getString(R.string.min)));
        }else {
            tv.setVisibility(View.GONE);
        }
    }

    public static Calendar getTimeFromText(CharSequence timeStr, String remTime, Date date){
        Calendar c = Calendar.getInstance();
        Date d = new Date();
        try {
            d = new Date(Objects.requireNonNull(new SimpleDateFormat("HH:mm", Locale.getDefault()).parse((String) timeStr)).getTime());
        } catch (ParseException e) {
            e.printStackTrace();
        }

        c.set(date.getYear(), date.getMonth(), date.getDate(), d.getHours(), d.getMinutes());

        return c;
    }  //todo прибавлять сутки если осталось больше 24
}

File: app/src/main/java/com/example/Sleepy/shared/MyTimer.kt
package com.example.Sleepy.shared

import android.content.Context
import com.example.Sleepy.R
import android.widget.TimePicker
import android.content.Intent
import android.provider.AlarmClock
import android.view.View
import com.google.android.material.snackbar.Snackbar
import android.widget.TextView
import java.lang.Exception
import java.util.*

class MyTimer {
    companion object{
        fun calcRemainingTimeMinute(date: Date, context: Context): String {
            val remTime: String
            val timeUp = date.time
            val diff = timeUp - Date().time
            val diffMinutes = diff / (60 * 1000) % 60
            val diffHours = diff / (60 * 60 * 1000)
            remTime = when {
                diffMinutes == 0L -> "" + diffHours + context.getString(R.string.hours)
                diffHours == 0L -> "" + diffMinutes + context.getString(R.string.minute)
                else -> "" + diffHours + context.getString(R.string.hours) + diffMinutes + context.getString(R.string.minute)
            }
            return remTime
        }

        fun getFormatTime(minute: Long, context: Context): String {
            val remTime: String
            val h: Long = minute / 60
            val m: Long = minute - h * 60
            remTime = when {
                m == 0L ->  "" + h + context.getString(R.string.hours)
                h != 0L && m != 0L -> "" + h + context.getString(R.string.hours) + m + context.getString(R.string.minute)
                else -> "" + m + context.getString(R.string.minute)
            }
            return remTime
        }

        fun clearTime(tp: TimePicker, context: Context) {
            MyVibrator.vibrate(30, context)
            tp.hour = Calendar.getInstance()[Calendar.HOUR_OF_DAY]
            tp.minute = Calendar.getInstance()[Calendar.MINUTE]
        }

        fun getCurrentTime(tp: TimePicker): Calendar {
            val timeAlarm = Calendar.getInstance()
            timeAlarm[Calendar.MINUTE] = tp.minute
            timeAlarm[Calendar.HOUR_OF_DAY] = tp.hour
            return timeAlarm
        }

        fun setAlarmInApp(time: Calendar, context: Context, view: View) {
            try {
                context.startActivity(
                    Intent(AlarmClock.ACTION_SET_ALARM)
                        .putExtra(AlarmClock.EXTRA_MESSAGE, "New Alarm")
                        .putExtra(AlarmClock.EXTRA_HOUR, time[Calendar.HOUR_OF_DAY])
                        .putExtra(AlarmClock.EXTRA_MINUTES, time[Calendar.MINUTE])
                )
            } catch (ex: Exception) {
                Snackbar.make(
                    view,
                    context.getString(R.string.error_open_alarm),
                    Snackbar.LENGTH_LONG
                ).show()
            }
        }

        fun getAsleepText(asleepTime: Int, tv: TextView, context: Context) {
            if (asleepTime != 0) {
                tv.visibility = View.VISIBLE
                tv.text = context.getString(R.string.calculate_with_sleep_time, asleepTime)
            } else {
                tv.visibility = View.GONE
            }
        }
    }
}